File: javascript/05-zombie-code/test/campaign_sender.test.js
import assert from 'node:assert';
import test from 'node:test';
import { CampaignSender } from '../src/campaign_sender.js';

test('counts EU customers with opt-in', () => {
  const sender = new CampaignSender();
  const result = sender.send_campaign([{ active: true, region: 'EU', gdprOptIn: true }], 'Hi');
  assert.strictEqual(result.sent, 1);
});

test('skips EU customers without opt-in', () => {
  const sender = new CampaignSender();
  const result = sender.send_campaign([{ active: true, region: 'EU', gdprOptIn: false }], 'Hi');
  assert.strictEqual(result.sent, 0);
});

test('counts non-EU active customers', () => {
  const sender = new CampaignSender();
  const result = sender.send_campaign([{ active: true, region: 'US' }], 'Hi');
  assert.strictEqual(result.sent, 1);
});

test('skips inactive customers', () => {
  const sender = new CampaignSender();
  const result = sender.send_campaign([{ active: false, region: 'US' }], 'Hi');
  assert.strictEqual(result.sent, 0);
  assert.strictEqual(result.skipped, 1);
});

test('handles mixed customers', () => {
  const sender = new CampaignSender();
  const result = sender.send_campaign([
    { active: true, region: 'EU', gdprOptIn: true },
    { active: true, region: 'EU', gdprOptIn: false },
    { active: true, region: 'US' },
    { active: false, region: 'US' }
  ], 'Hi');
  assert.strictEqual(result.sent, 2);
  assert.strictEqual(result.skipped, 2);
});

test('skips unsubscribed customers', () => {
  const sender = new CampaignSender();
  const result = sender.send_campaign([{ active: true, region: 'US', unsubscribed: true }], 'Hi');
  assert.strictEqual(result.sent, 0);
  assert.strictEqual(result.skipped, 1);
});

test('dry run never sends but keeps skip count', () => {
  const sender = new CampaignSender();
  const result = sender.send_campaign([
    { active: true, region: 'US' },
    { active: false, region: 'US' }
  ], '__dry_run__');
  assert.strictEqual(result.sent, 0);
  assert.strictEqual(result.skipped, 1);
});

test('returns message in result', () => {
  const sender = new CampaignSender();
  const result = sender.send_campaign([], 'Hello World');
  assert.strictEqual(result.message, 'Hello World');
});

test('returns timestamp in result', () => {
  const sender = new CampaignSender();
  const result = sender.send_campaign([], 'Hi');
  assert.ok(typeof result.timestamp === 'number');
});

File: python/05-zombie-code/test/campaign_sender_test.py
from src.campaign_sender import CampaignSender

def test_eu_opt_in():
    sender = CampaignSender()
    result = sender.send_campaign([{"active": True, "region": "EU", "gdprOptIn": True}], "Hi")
    assert result["sent"] == 1

def test_eu_no_opt_in():
    sender = CampaignSender()
    result = sender.send_campaign([{"active": True, "region": "EU", "gdprOptIn": False}], "Hi")
    assert result["sent"] == 0

def test_non_eu_active():
    sender = CampaignSender()
    result = sender.send_campaign([{"active": True, "region": "US"}], "Hi")
    assert result["sent"] == 1

def test_inactive_customer():
    sender = CampaignSender()
    result = sender.send_campaign([{"active": False, "region": "US"}], "Hi")
    assert result["sent"] == 0
    assert result["skipped"] == 1

def test_mixed_customers():
    sender = CampaignSender()
    customers = [
        {"active": True, "region": "EU", "gdprOptIn": True},
        {"active": True, "region": "EU", "gdprOptIn": False},
        {"active": True, "region": "US"},
        {"active": False, "region": "US"}
    ]
    result = sender.send_campaign(customers, "Hi")
    assert result["sent"] == 2
    assert result["skipped"] == 2

def test_unsubscribed_customer():
    sender = CampaignSender()
    result = sender.send_campaign([{"active": True, "region": "US", "unsubscribed": True}], "Hi")
    assert result["sent"] == 0
    assert result["skipped"] == 1

def test_dry_run_never_sends_but_keeps_skip_count():
    sender = CampaignSender()
    result = sender.send_campaign([
        {"active": True, "region": "US"},
        {"active": False, "region": "US"}
    ], "__dry_run__")
    assert result["sent"] == 0
    assert result["skipped"] == 1

def test_message_in_result():
    sender = CampaignSender()
    result = sender.send_campaign([], "Hello World")
    assert result["message"] == "Hello World"

def test_timestamp_is_number():
    sender = CampaignSender()
    result = sender.send_campaign([], "Hi")
    assert isinstance(result["timestamp"], int)
